helm: update docs to clarify chart maintenance requirements (#1658)
kubernetes-sigs/cloud-provider-azure: helm/cloud-provider-azure/README.md

# Maintaining the Repo

Whenever changes have been made to the `cloud-provider-azure` Helm chart, a new version of the chart should be released. First, pick an appropriate next, higher version and update the `version` property in `helm/cloud-provider-azure/Chart.yaml`. Then, package the entire set of changes to the chart into a new repo version:

From the git root:

```bash
$ make update-helm
Successfully packaged chart and saved it to: helm/repo/cloud-provider-azure-1.23.11.tgz
```

Changes to the chart should *always* include a new version, and then an update to the Helm repo as described above.

If changes to the chart don't include an update to the Helm repo as described above, then the PR CI job will fail, preventing a merge to `master`.

To check if your Helm chart changes will pass CI, you can manually run the verification script:

```bash
$ hack/verify-helm-repo.sh

> ++ dirname hack/verify-helm-repo.sh

Reviewer: is it possible to move the script into a file?

Reviewer 2: @nilo19 this is actually not a script. Just run the command `hack/verify-helm-repo.sh` and its output.

Reviewer 2: @jackfrancis I think you ran the old script and could you please rerun with the latest one? I just updated it yesterday.
https://github.com/kubernetes-sigs/cloud-provider-azure/pull/1656#event-6556909503

Author: updated output rebasing on top of your script change, thanks!

+ rm -Rf hack/../chart_verify
+ mkdir hack/../chart_verify
+ tar -xf hack/../helm/repo/cloud-provider-azure-1.23.11.tgz -C chart_verify
+ diff -r hack/../chart_verify/cloud-provider-azure/ hack/../helm/cloud-provider-azure/ --exclude README.md
+ rm -Rf hack/../chart_verify
+ exit 0
$ echo $?
0
```

## How to pick an appropriate new version when updating the Helm chart

By convention, we pin the Helm chart version to the most recent release of cloud-provider-azure supporting the most recent release of Kubernetes. For example, at the time of this writing, the most recent version of Kubernetes supported by cloud-provider-azure is 1.23, and the most recent release of cloud-provider-azure is `v1.23.11`. When the next version of cloud-provider-azure is released (either `v1.23.12`, or `v1.24.0` in the event that the next release includes support for a new minor release of Kubernetes) we would update the Helm chart to refer to the newly released, versioned image references of `cloud-controller-manager` and `cloud-node-manager`, and then update the Helm chart version to the same version of of the new release (e.g., either `v1.23.12` or `1.24.0`).

> If, however, changes to the chart are made independently of a new release of cloud-provider-azure, we can follow the semantic versioning convention of adding `suffix` data to the exiting release. For example, if the current release of the Helm chart is `v1.23.11` and we want to add additional configuration capabilities to `values.yaml`, we may do so, and then update the version of the chart to `v1.23.11+20220505` following the formula `<existing release>+<current date in format YYYYMMDD>`.

Reviewer 2: ```suggestion
If, however, changes to the chart are made independently of a new release of cloud-provider-azure, we can follow the semantic versioning convention of adding `suffix` data to the existing release. For example, if the current release of the Helm chart is `v1.23.11` and we want to add additional configuration capabilities to `values.yaml`, we may do so, and then update the version of the chart to `v1.23.11+20220505` following the formula `<existing release>+<current date in format YYYYMMDD>`.
```

Author: thanks!

```bash
$ make update-helm
Successfully packaged chart and saved it to: helm/repo/cloud-provider-azure-1.23.11.tgz
```

Changes to the chart should *always* include a new version, and then an update to the helm repo as described above.

> If changes to the chart don't include an update to the helm repo as described above, then the PR CI job will fail, preventing a merge to `master`.

Reviewer 3: ```suggestion
If changes to the chart don't include an update to the Helm repo as described above, then the PR CI job will fail, preventing a merge to `master`.
```

To check if your helm changes will pass CI, you can manually run the verification script:

```bash
$ hack/verify-helm-repo.sh
++ dirname hack/verify-helm-repo.sh
+ REPO_ROOT=hack/..
+ for chart in '"cloud-provider-azure"'
++ head -n 1
++ ls -t hack/../helm/repo/cloud-provider-azure-1.23.10.tgz hack/../helm/repo/cloud-provider-azure-1.23.11.tgz hack/../helm/repo/cloud-provider-azure-1.23.8.tgz
+ LATEST_CHART=hack/../helm/repo/cloud-provider-azure-1.23.11.tgz
++ echo hack/../helm/repo/cloud-provider-azure-1.23.11.tgz
++ grep -Eoq '[0-9]+.[0-9]+.[0-9]+'
+ LATEST_VERSION=
+ MATCH_STRING='version: '
+ grep -q 'version: ' hack/../helm/cloud-provider-azure/Chart.yaml
+ rm -Rf hack/../chart_verify
+ mkdir hack/../chart_verify
+ tar -xf hack/../helm/repo/cloud-provider-azure-1.23.11.tgz -C chart_verify
+ diff -r hack/../chart_verify/cloud-provider-azure/ hack/../helm/cloud-provider-azure/ --exclude README.md
+ rm -Rf hack/../chart_verify
+ exit 0
$ echo $?
0
```

> ## How to pick an appropriate new version when updating the helm chart

Reviewer 3: ```suggestion
## How to pick an appropriate new version when updating the Helm chart
```

> By convention, we pin the helm chart version to the most recent release of cloud-provider-azure supporting the most recent release of Kubernetes. For example, at the time of this writing, the most recent version of Kubernetes supported by cloud-provider-azure is 1.23, and the most recent release of cloud-provider-azure is `v1.23.11`. When the next version of cloud-provider-azure is released (either `v1.23.12`, or `v1.24.0` in the event that the next release includes support for a new minor release of Kubernetes) we would update the helm chart to refer to the newly released, versioned image references of `cloud-controller-manager` and `cloud-node-manager`, and then update the helm chart version to the same version of of the new release (e.g., either `v1.23.12` or `1.24.0`).

Reviewer 3: ```suggestion
By convention, we pin the Helm chart version to the most recent release of cloud-provider-azure supporting the most recent release of Kubernetes. For example, at the time of this writing, the most recent version of Kubernetes supported by cloud-provider-azure is 1.23, and the most recent release of cloud-provider-azure is `v1.23.11`. When the next version of cloud-provider-azure is released (either `v1.23.12`, or `v1.24.0` in the event that the next release includes support for a new minor release of Kubernetes) we would update the Helm chart to refer to the newly released, versioned image references of `cloud-controller-manager` and `cloud-node-manager`, and then update the Helm chart version to the same version of of the new release (e.g., either `v1.23.12` or `1.24.0`).
```

> If, however, changes to the chart are made independent of a new release of cloud-provider-azure, we can follow the semver convention of adding `suffix` data to the exiting release. For example, if the current release of the helm chart is `v1.23.11` and we want to add additional configuration capabilities to `values.yaml`, we may do so, and then update the version of the chart to `v1.23.11+20220505` following the formula `<existing release>+<current date in format YYYYMMDD>`.

Reviewer 3: ```suggestion
If, however, changes to the chart are made independently of a new release of cloud-provider-azure, we can follow the semantic versioning convention of adding `suffix` data to the exiting release. For example, if the current release of the Helm chart is `v1.23.11` and we want to add additional configuration capabilities to `values.yaml`, we may do so, and then update the version of the chart to `v1.23.11+20220505` following the formula `<existing release>+<current date in format YYYYMMDD>`.
```

> As described [above](#maintaining-the-repo), always run `make update-helm` after making changes to the helm chart, and updating the version.

Reviewer 3: ```suggestion
As described [above](#maintaining-the-repo), always run `make update-helm` after making changes to the Helm chart, and updating the version.
```
